### PR TITLE
Use intrinsic instead of inline asm for lds barrier

### PR DESCRIPTION
**Pull request by jungpark-mlir**
Let's see how it works,
- compare the performance as is.
- We can build llvm (tip) separately, can we run tests with it?

### CODE REVIEW

**Comment by whchung** [comment]
@yiqian1 Wondering how do we set up a performance run for this PR? We want to understand the performance characteristics switching from emitting inline assembly to using LLVM intrinsics.

**Comment by jerryyin** [comment]
With [this PR](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/376), it should allow us to execute any independent parallel stage if we clicked on the check of that stage.

Right now this is what we have:
Screenshot from 2021-09-13 11-24-20

After the PR is merged, we can click on `perfTest` toggle and just execute that stage.

**Comment by jerryyin** [comment]
~~@jungpark-mlir  Could you do a rebase to catch up with the tip? I can show you how to run a perf regression test.~~

Performance run, when finished, will be in [here](http://10.236.106.97:8080/blue/organizations/jenkins/mlir/detail/PR-371/4/pipeline/10).

**Comment by jerryyin** [comment]
According to the performance report [here](http://10.236.106.97:8080/job/mlir/job/PR-371/4/Performance_20report/), it stays even with past mlir performance.

File: external/llvm-project/mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td [hunk 1]
```diff
@@ -106,18 +106,11 @@ def ROCDL_BarrierOp : ROCDL_Op<"barrier"> {
...
-        function_type,                       // FunctionType *Ty,
-        "s_waitcnt lgkmcnt(0) \n s_barrier", // const std::string &AsmString,
-        "",                                  // const std::string &Constraints,
-        true,                                // bool hasSideEffects,
-        false,                               // bool isAlignStack,
-        llvm::InlineAsm::AD_ATT              // AsmDialect asmDialect
-        );
-    builder.CreateCall(function_type, inlineAsm);
+    //    builder.CreateFence(llvm::AtomicOrdering::Release,
+    //                       llvmContext.getOrInsertSyncScopeID("workgroup"));
```
Comment (whchung) [inline external/llvm-project/mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td]: @jungpark-mlir now the PR has got a performance run and looks fine, we should consider land it.

do you think we should use LLVM fence with workgroup synscope, like what you commented here, or directly create intrinsic calls, like what's being implemented below?

**Comment by jerryyin** [comment]
Sorry, taking a look at the performance report, it doesn't seem to compare with the most recent MLIR perf. (I have no idea where it get that past mlir perf from.) I also pulled up the most recent nightly perf report, [here](http://10.236.106.97:8080/job/mlir-nightly-all/99/Performance_20report/). Comparing between the two visually I feel like this's slight amount of degrade, probably 2% ish. Considering that the two run may run on different machines, the conclusion of perf stays even seem to stand.

**Comment by whchung** [comment]
> Sorry, taking a look at the performance report, it doesn't seem to compare with the most recent MLIR perf. (I have no idea where it get that past mlir perf from.) I also pulled up the most recent nightly perf report, [here](http://10.236.106.97:8080/job/mlir-nightly-all/99/Performance_20report/). Comparing between the two visually I feel like this's slight amount of degrade, probably 2% ish. Considering that the two run may run on different machines, the conclusion of perf stays even seem to stand.

I'm willing to spend 2% over undesirable inline asm, which might easily bite us in other places.

I'd really be interested in observing the perf numbers after we take @sjw36 's PR #378 .

File: external/llvm-project/mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td (hunk 1 above)
Comment (jungpark-mlir) [reply external/llvm-project/mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td]: That's good point, I felt compiler team prefers to use fence. I'll change to it, generate instruction should be identical and will not change the performance.

**Comment by krzysz00** [comment]
@jerryyin What's going on is that the performance report generator tries to fetch the artifacts from the previous successful run of the same job, which, in the case of a PR, means it fails to actually fetch any performance reports. We may want to have a PR-specific limited performance tests (maybe it just looks at a subset of the kernels) for this sort of thing.

**Review by whchung — APPROVED** [review]

**Pull request merged**